sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefService.java

```diff
@@ -196,43 +191,170 @@ public static List<TExtRef> filterDuplicatedExtRefs(List<TExtRef> tExtRefs) {
+                && extRef.isSetPDO() && extRef.getPDO().equals(tChannel.getDOName());
+    }
+
+    /**
+     * Verify whether the IED satisfies the EPF channel for the private element `TCompasICDHeader`
+     * @param iedAdapter IEDAdapter
+     * @param channel TChannel
+     * @return true if the TCompasICDHeader matches the EPF channel
+     */
+    private static boolean isIcdHeaderMatchLDEPFChannel(IEDAdapter iedAdapter, TChannel channel) {
+        return iedAdapter.getCompasICDHeader()
+                .map(compasICDHeader -> compasICDHeader.getIEDType().value().equals(channel.getIEDType())
+                        && compasICDHeader.getIEDredundancy().value().equals(channel.getIEDRedundancy().value())
+                        && compasICDHeader.getIEDSystemVersioninstance().toString().equals(channel.getIEDSystemVersionInstance()))
+                .orElse(false);
+    }
+
+    /**
+     * Provides Active LDevice according to EPF channel's inst attribute
+     * @param iedAdapter IEDAdapter
+     * @param channel TChannel
+     * @return LDeviceAdapter object that matches the EPF channel
+     */
+    private static Optional<LDeviceAdapter> getActiveSourceLDeviceByLDEPFChannel(IEDAdapter iedAdapter, TChannel channel) {
+        return iedAdapter.findLDeviceAdapterByLdInst(channel.getLDInst())
+                .filter(lDeviceAdapter -> lDeviceAdapter.getLDeviceStatus()
+                        .map(status -> status.equals(LdeviceStatus.ON.getValue()))
+                        .orElse(false))
+                .stream().findFirst();
+    }
+
+    /**
+     * Provides Active LN Object that satisfies the EPF channel attributes (lnClass, lnInst, prefix)
+     * @param lDeviceAdapter LDeviceAdapter
+     * @param channel TChannel
+     * @return AbstractLNAdapter object that matches the EPF channel
+     */
+    private static Optional<AbstractLNAdapter<?>> getActiveLNodeSourceByLDEPFChannel(LDeviceAdapter lDeviceAdapter, TChannel channel) {
+        return lDeviceAdapter.getLNAdaptersIncludingLN0().stream()
+                .filter(lnAdapter -> lnAdapter.getLNClass().equals(channel.getLNClass())
+                        && lnAdapter.getLNInst().equals(channel.getLNInst())
+                        && Utils.equalsOrBothBlank(channel.getLNPrefix(), lnAdapter.getPrefix()))
+                .findFirst()
+                .filter(lnAdapter -> lnAdapter.getDaiModStValValue()
+                        .map(status -> status.equals(LdeviceStatus.ON.getValue()))
+                        .orElse(true));
+    }
+
+    /**
+     *  Verify whether the LN satisfies the EPF channel parameters for Data Type Template elements.
+     * @param lnAdapter AbstractLNAdapter
+     * @param channel TChannel
+     * @return true if the LN matches the EPF channel
+     */
+    private static boolean isValidDataTypeTemplate(AbstractLNAdapter<?> lnAdapter, TChannel channel) {
+        if(StringUtils.isEmpty(channel.getDOName()) || StringUtils.isBlank(channel.getDOName())){
```

Comment:
just use isBlank (as the javadoc say "_**Checks if a CharSequence is empty (""), null or whitespace only.**_")

Reply:
As TChannel was defined with XSD, we can use `!channel.isSetDOName()` instead of isEmpty

Reply:
Done